File: src/personal_data_warehouse_apple_notes/sync.py
from __future__ import annotations

from collections.abc import Callable
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
import hashlib
from html import escape
import json
import tempfile
import threading

from personal_data_warehouse_apple_notes.scanner import (
    AppleNote,
    AppleNoteAttachment,
    parse_datetime,
    scan_apple_notes_store,
    snapshot_apple_notes_store,
)
from personal_data_warehouse_apple_notes.state import AppleNotesUploadState


@dataclass(frozen=True)
class AppleNotesUploadSummary:
    notes_seen: int
    notes_selected: int
    notes_skipped: int
    notes_deleted: int
    revisions_uploaded: int
    metadata_uploaded: int
    body_html_uploaded: int
    attachments_seen: int
    attachments_uploaded: int
    attachments_missing: int
    notes_deferred: int = 0


@dataclass(frozen=True)
class AppleNoteRevision:
    note: AppleNote
    revision_id: str
    fingerprint: str
    is_tombstone: bool = False


@dataclass(frozen=True)
class AppleNoteUploadResult:
    revision_uploaded: int
    metadata_uploaded: int
    html_uploaded: int
    attachments_uploaded: int
    attachments_missing: int


class AppleNotesUploadRunner:
    def __init__(
        self,
        *,
        account: str,
        store_path: Path | str,
        ingest_client,
        logger,
        upload_state: AppleNotesUploadState | None = None,
        now: Callable[[], datetime] | None = None,
        mode: str = "incremental",
        before_upload_check: Callable[[], str | None] | None = None,
        limit: int | None = None,
        workers: int = 1,
        state_save_callback: Callable[[], None] | None = None,
    ) -> None:
        # Uploads go through the app's ingest endpoints only.
        if ingest_client is None:
            raise ValueError("ingest_client is required")
        if mode not in {"full", "incremental"}:
            raise ValueError("mode must be 'full' or 'incremental'")
        if limit is not None and limit < 0:
            raise ValueError("limit must be nonnegative")
        self._account = account
        self._store_path = Path(store_path).expanduser()
        self._ingest_client = ingest_client
        self._state_lock = threading.Lock()
        self._logger = logger
        self._upload_state = upload_state
        self._now = now or (lambda: datetime.now(tz=UTC))
        self._mode = mode
        self._before_upload_check = before_upload_check
        self._limit = limit
        self._workers = max(1, workers)
        self._state_save_callback = state_save_callback

    def sync(self) -> AppleNotesUploadSummary:
        self._logger.info("Snapshotting Apple Notes store at %s", self._store_path)
        with tempfile.TemporaryDirectory(prefix="pdw-apple-notes-") as temp_dir:
            snapshot_path = snapshot_apple_notes_store(self._store_path, temp_dir)
            notes = scan_apple_notes_store(snapshot_path, attachments_root=self._store_path.parent)
        notes_by_id = {note.note_id: note for note in notes}
        revisions = [revision_from_note(note) for note in notes]
        tombstones = self._deleted_note_revisions(current_note_ids=set(notes_by_id))
        revisions.extend(tombstones)

        notes_seen = len(notes)
        attachments_seen = sum(len(note.attachments) for note in notes)
        attachments_missing = sum(1 for note in notes for attachment in note.attachments if attachment.is_missing)
        selected: list[AppleNoteRevision] = []
        skipped = 0
        for revision in revisions:
            entry = self._upload_state.entry_for(revision.note.note_id) if self._upload_state else None
            if (
                self._mode == "incremental"
                and entry is not None
                and entry.complete
                and entry.fingerprint == revision.fingerprint
                and entry.is_deleted == revision.note.is_deleted
            ):
                skipped += 1
                continue
            selected.append(revision)

        deferred = 0
        if self._limit is not None and len(selected) > self._limit:
            deferred = len(selected) - self._limit
            selected = selected[: self._limit]

        if selected and self._before_upload_check is not None:
            skip_reason = self._before_upload_check()
            if skip_reason:
                self._logger.warning("Skipping Apple Notes upload: %s", skip_reason)
                return AppleNotesUploadSummary(
                    notes_seen=notes_seen,
                    notes_selected=len(selected),
                    notes_skipped=skipped,
                    notes_deleted=len(tombstones),
                    revisions_uploaded=0,
                    metadata_uploaded=0,
                    body_html_uploaded=0,
                    attachments_seen=attachments_seen,
                    attachments_uploaded=0,
                    attachments_missing=attachments_missing,
                    notes_deferred=len(selected) + deferred,
                )

        self._logger.info("Uploading with %s worker(s)", self._workers)
        if self._workers == 1 or len(selected) <= 1:
            results = [
                self._sync_revision(index=index, total=len(selected), revision=revision)
                for index, revision in enumerate(selected, start=1)
            ]
        else:
            with ThreadPoolExecutor(max_workers=self._workers, thread_name_prefix="apple-notes") as executor:
                futures = [
                    executor.submit(self._sync_revision, index=index, total=len(selected), revision=revision)
                    for index, revision in enumerate(selected, start=1)
                ]
                results = [future.result() for future in as_completed(futures)]

        summary = AppleNotesUploadSummary(
            notes_seen=notes_seen,
            notes_selected=len(selected),
            notes_skipped=skipped,
            notes_deleted=len(tombstones),
            revisions_uploaded=sum(result.revision_uploaded for result in results),
            metadata_uploaded=sum(result.metadata_uploaded for result in results),
            body_html_uploaded=sum(result.html_uploaded for result in results),
            attachments_seen=attachments_seen,
            attachments_uploaded=sum(result.attachments_uploaded for result in results),
            attachments_missing=attachments_missing,
            notes_deferred=deferred,
        )
        self._logger.info(
            "Apple Notes upload summary: seen=%s selected=%s skipped=%s deferred=%s revisions=%s metadata=%s html=%s attachments=%s missing=%s deleted=%s",
            summary.notes_seen,
            summary.notes_selected,
            summary.notes_skipped,
            summary.notes_deferred,
            summary.revisions_uploaded,
            summary.metadata_uploaded,
            summary.body_html_uploaded,
            summary.attachments_uploaded,
            summary.attachments_missing,
            summary.notes_deleted,
        )
        return summary

    def _deleted_note_revisions(self, *, current_note_ids: set[str]) -> list[AppleNoteRevision]:
        if self._upload_state is None:
            return []
        revisions: list[AppleNoteRevision] = []
        deleted_at = self._now()
        for note_id, entry in sorted(self._upload_state.entries.items()):
            if note_id in current_note_ids or entry.is_deleted:
                continue
            note = AppleNote(
                note_id=note_id,
                title=entry.title,
                folder_id="",
                folder_path="",
                apple_account_id="",
                apple_account_name="",
                created_at=parse_datetime(entry.modified_at),
                modified_at=deleted_at,
                body_text="",
                body_html="",
                body_markdown="",
                attachments=(),
                is_deleted=True,
                raw={"tombstone_from_revision_id": entry.revision_id},
            )
            revisions.append(revision_from_note(note))
        return revisions

    def _sync_revision(self, *, index: int, total: int, revision: AppleNoteRevision) -> AppleNoteUploadResult:
        try:
            self._logger.info(
                "[%s/%s] upload Apple Note %s revision %s -> app",
                index,
                total,
                revision.note.note_id,
                short_sha256(revision.revision_id),
            )
            result = self._upload_revision(revision)
            self._mark_success(revision=revision, result=result)
            return result
        except Exception as exc:
            self._mark_failure(revision=revision, error=str(exc))
            raise

    def _mark_success(self, *, revision: AppleNoteRevision, result: AppleNoteUploadResult) -> None:
        if self._upload_state is None:
            return
        with self._state_lock:
            self._upload_state.mark_success(
                note_id=revision.note.note_id,
                fingerprint=revision.fingerprint,
                revision_id=revision.revision_id,
                title=revision.note.title,
                modified_at=revision.note.modified_at,
                is_deleted=revision.note.is_deleted,
                metadata_uploaded=True,
                html_uploaded=bool(revision.note.is_deleted or result.html_uploaded),
                attachments_uploaded=True,
                now=self._now(),
            )
            if self._state_save_callback is not None:
                self._state_save_callback()

    def _mark_failure(self, *, revision: AppleNoteRevision, error: str) -> None:
        if self._upload_state is None:
            return
        with self._state_lock:
            self._upload_state.mark_failure(
                note_id=revision.note.note_id,
                error=error,
                now=self._now(),
            )
            if self._state_save_callback is not None:
                self._state_save_callback()


    def _upload_revision(self, revision: AppleNoteRevision) -> AppleNoteUploadResult:
        # The app owns object keys, kinds, and pdw_* tags; we send only the
        # revision's domain fields and bytes.
        payload = metadata_payload(account=self._account, revision=revision, exported_at=self._now())
        modified_at = revision.note.modified_at.astimezone(UTC).isoformat()
        note_id = revision.note.note_id
        revision_id = revision.revision_id
        html_uploaded = 0
        attachment_uploads = 0
        attachment_missing = sum(1 for attachment in revision.note.attachments if attachment.is_missing)

        if not revision.note.is_deleted:
            html = revision.note.body_html or html_document_for_note(revision.note)
            if html:
                self._ingest_client.upload_apple_notes_body(
                    html.encode("utf-8"),
                    note_id=note_id,
                    revision_id=revision_id,
                    modified_at=modified_at,
                )
                html_uploaded = 1
            for attachment in revision.note.attachments:
                if attachment.is_missing or attachment.path is None or not attachment.content_sha256:
                    continue
                self._ingest_client.upload_apple_notes_attachment(
                    Path(attachment.path).read_bytes(),
                    note_id=note_id,
                    revision_id=revision_id,
                    modified_at=modified_at,
                    attachment_id=attachment.attachment_id,
                    filename=attachment.filename,
                    content_type=attachment.content_type,
                )
                attachment_uploads += 1

        self._ingest_client.upload_apple_notes_revision(
            payload,
            note_id=note_id,
            revision_id=revision_id,
            modified_at=modified_at,
            note_content_sha256=revision.fingerprint,
        )
        return AppleNoteUploadResult(
            revision_uploaded=1,
            metadata_uploaded=1,
            html_uploaded=html_uploaded,
            attachments_uploaded=attachment_uploads,
            attachments_missing=attachment_missing,
        )


def revision_from_note(note: AppleNote) -> AppleNoteRevision:
    fingerprint_payload = note_payload(note)
    fingerprint = json_sha256(fingerprint_payload)
    return AppleNoteRevision(note=note, revision_id=fingerprint, fingerprint=fingerprint, is_tombstone=note.is_deleted)


def note_payload(note: AppleNote) -> dict[str, object]:
    return {
        "note_id": note.note_id,
        "title": note.title,
        "folder_id": note.folder_id,
        "folder_path": note.folder_path,
        "apple_account_id": note.apple_account_id,
        "apple_account_name": note.apple_account_name,
        "created_at": note.created_at.astimezone(UTC).isoformat(),
        "modified_at": note.modified_at.astimezone(UTC).isoformat(),
        "body_text": note.body_text,
        "body_html": note.body_html,
        "body_markdown": note.body_markdown,
        "attachments": [attachment_payload(attachment) for attachment in note.attachments],
        "is_deleted": note.is_deleted,
        "raw": note.raw or {},
    }


def attachment_payload(attachment: AppleNoteAttachment) -> dict[str, object]:
    return {
        "attachment_id": attachment.attachment_id,
        "note_id": attachment.note_id,
        "filename": attachment.filename,
        "content_type": attachment.content_type,
        "size_bytes": attachment.size_bytes,
        "content_sha256": attachment.content_sha256,
        "is_missing": attachment.is_missing,
        "error": attachment.error,
        "raw": attachment.raw or {},
    }


def metadata_payload(*, account: str, revision: AppleNoteRevision, exported_at: datetime) -> dict[str, object]:
    note = note_payload(revision.note)
    note["revision_id"] = revision.revision_id
    note["content_sha256"] = revision.fingerprint
    return {
        "schema_version": 1,
        "source": "apple_notes",
        "account": account,
        "exported_at": exported_at.astimezone(UTC).isoformat(),
        "note": note,
    }


def html_document_for_note(note: AppleNote) -> str:
    title = escape(note.title or note.note_id)
    body = note.body_html or ("<pre>" + escape(note.body_text) + "</pre>")
    return f"<!doctype html><html><head><meta charset=\"utf-8\"><title>{title}</title></head><body>{body}</body></html>"


def json_sha256(payload: dict[str, object]) -> str:
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":"), default=str).encode("utf-8")
    return hashlib.sha256(encoded).hexdigest()


def short_sha256(value: str) -> str:
    return value[:12]

File: app/internal/uploaders/applenotes/sync.go
package applenotes

import (
	"errors"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/zachlatta/personal-data-warehouse/app/internal/ingestclient"
	"github.com/zachlatta/personal-data-warehouse/app/internal/uploaders/common"
)

// Summary is the run's counts.
type Summary struct {
	NotesSeen           int
	NotesSelected       int
	NotesSkipped        int
	NotesDeleted        int
	RevisionsUploaded   int
	MetadataUploaded    int
	BodyHTMLUploaded    int
	AttachmentsSeen     int
	AttachmentsUploaded int
	AttachmentsMissing  int
	NotesDeferred       int
}

// Revision is one note snapshot to upload; its fingerprint is the revision id.
type Revision struct {
	Note        Note
	RevisionID  string
	Fingerprint string
	IsTombstone bool
}

// Uploader is what the runner needs from the ingest client.
type Uploader interface {
	UploadAppleNotesBody(html []byte, noteID, revisionID, modifiedAt string) (ingestclient.StoredObject, error)
	UploadAppleNotesAttachment(content []byte, a ingestclient.AppleNotesAttachment) (ingestclient.StoredObject, error)
	UploadAppleNotesRevision(payload map[string]any, noteID, revisionID, modifiedAt, noteContentSHA256 string) (ingestclient.StoredObject, error)
}

// Runner uploads changed notes and tombstones for vanished ones.
type Runner struct {
	Account           string
	StorePath         string
	Client            Uploader
	Logger            common.Logger
	State             *State
	Now               func() time.Time
	Mode              string
	BeforeUploadCheck func() string
	Limit             int
	Workers           int
	SaveState         func() // called after every state change

	stateMu sync.Mutex
}

type uploadResult struct {
	revisionUploaded    int
	metadataUploaded    int
	htmlUploaded        int
	attachmentsUploaded int
	attachmentsMissing  int
}

// Sync runs one pass.
func (r *Runner) Sync() (Summary, error) {
	if r.Client == nil {
		return Summary{}, errors.New("ingest client is required")
	}
	if r.Mode == "" {
		r.Mode = "incremental"
	}
	if r.Mode != "incremental" && r.Mode != "full" {
		return Summary{}, errors.New("mode must be 'full' or 'incremental'")
	}
	if r.Now == nil {
		r.Now = func() time.Time { return time.Now().UTC() }
	}
	if r.Workers < 1 {
		r.Workers = 1
	}
	storePath := common.ExpandUser(r.StorePath)
	r.Logger.Infof("Snapshotting Apple Notes store at %s", storePath)
	tempDir, err := os.MkdirTemp("", "pdw-apple-notes-")
	if err != nil {
		return Summary{}, err
	}
	snapshot, err := SnapshotStore(storePath, tempDir)
	if err != nil {
		os.RemoveAll(tempDir)
		return Summary{}, err
	}
	notes, err := Scan(snapshot, filepath.Dir(storePath))
	os.RemoveAll(tempDir)
	if err != nil {
		return Summary{}, err
	}
	return r.syncNotes(notes)
}

func (r *Runner) syncNotes(notes []Note) (Summary, error) {
	if r.Workers < 1 {
		r.Workers = 1
	}
	if r.Mode == "" {
		r.Mode = "incremental"
	}
	if r.Now == nil {
		r.Now = func() time.Time { return time.Now().UTC() }
	}
	current := map[string]bool{}
	var revisions []Revision
	for _, note := range notes {
		current[note.NoteID] = true
		revisions = append(revisions, RevisionFromNote(note))
	}
	tombstones := r.deletedNoteRevisions(current)
	revisions = append(revisions, tombstones...)

	summary := Summary{NotesSeen: len(notes), NotesDeleted: len(tombstones)}
	for _, note := range notes {
		summary.AttachmentsSeen += len(note.Attachments)
		for _, attachment := range note.Attachments {
			if attachment.IsMissing {
				summary.AttachmentsMissing++
			}
		}
	}
	var selected []Revision
	for _, revision := range revisions {
		if r.Mode == "incremental" && r.State != nil {
			if entry, ok := r.State.EntryFor(revision.Note.NoteID); ok && entry.Complete() && entry.Fingerprint == revision.Fingerprint && entry.IsDeleted == revision.Note.IsDeleted {
				summary.NotesSkipped++
				continue
			}
		}
		selected = append(selected, revision)
	}
	deferred := 0
	if r.Limit > 0 && len(selected) > r.Limit {
		deferred = len(selected) - r.Limit
		selected = selected[:r.Limit]
	}
	summary.NotesSelected = len(selected)
	summary.NotesDeferred = deferred
	if len(selected) > 0 && r.BeforeUploadCheck != nil {
		if reason := r.BeforeUploadCheck(); reason != "" {
			r.Logger.Warningf("Skipping Apple Notes upload: %s", reason)
			summary.NotesDeferred = len(selected) + deferred
			return summary, nil
		}
	}
	r.Logger.Infof("Uploading with %d worker(s)", r.Workers)
	results := make([]uploadResult, len(selected))
	errs := make([]error, len(selected))
	var wg sync.WaitGroup
	semaphore := make(chan struct{}, r.Workers)
	for i, revision := range selected {
		wg.Add(1)
		semaphore <- struct{}{}
		go func(i int, revision Revision) {
			defer wg.Done()
			defer func() { <-semaphore }()
			results[i], errs[i] = r.syncRevision(i+1, len(selected), revision)
		}(i, revision)
	}
	wg.Wait()
	for _, err := range errs {
		if err != nil {
			return summary, err
		}
	}
	for _, result := range results {
		summary.RevisionsUploaded += result.revisionUploaded
		summary.MetadataUploaded += result.metadataUploaded
		summary.BodyHTMLUploaded += result.htmlUploaded
		summary.AttachmentsUploaded += result.attachmentsUploaded
	}
	r.Logger.Infof("Apple Notes upload summary: seen=%d selected=%d skipped=%d deferred=%d revisions=%d metadata=%d html=%d attachments=%d missing=%d deleted=%d",
		summary.NotesSeen, summary.NotesSelected, summary.NotesSkipped, summary.NotesDeferred, summary.RevisionsUploaded,
		summary.MetadataUploaded, summary.BodyHTMLUploaded, summary.AttachmentsUploaded, summary.AttachmentsMissing, summary.NotesDeleted)
	return summary, nil
}

func (r *Runner) deletedNoteRevisions(current map[string]bool) []Revision {
	if r.State == nil {
		return nil
	}
	var revisions []Revision
	deletedAt := r.Now()
	for _, noteID := range r.State.SortedNoteIDs() {
		entry := r.State.Entries[noteID]
		if current[noteID] || entry.IsDeleted {
			continue
		}
		note := Note{
			NoteID:     noteID,
			Title:      entry.Title,
			CreatedAt:  common.ParseISO(entry.ModifiedAt),
			ModifiedAt: deletedAt,
			IsDeleted:  true,
			Raw:        map[string]any{"tombstone_from_revision_id": entry.RevisionID},
		}
		revisions = append(revisions, RevisionFromNote(note))
	}
	return revisions
}

func (r *Runner) syncRevision(index, total int, revision Revision) (uploadResult, error) {
	r.Logger.Infof("[%d/%d] upload Apple Note %s revision %s -> app", index, total, revision.Note.NoteID, common.ShortSHA256(revision.RevisionID))
	result, err := r.uploadRevision(revision)
	if err != nil {
		r.markFailure(revision, err.Error())
		return result, err
	}
	r.markSuccess(revision, result)
	return result, nil
}

func (r *Runner) markSuccess(revision Revision, result uploadResult) {
	if r.State == nil {
		return
	}
	r.stateMu.Lock()
	defer r.stateMu.Unlock()
	r.State.MarkSuccess(revision.Note.NoteID, revision.Fingerprint, revision.RevisionID, revision.Note.Title, revision.Note.ModifiedAt,
		revision.Note.IsDeleted, true, revision.Note.IsDeleted || result.htmlUploaded > 0, true, r.Now())
	if r.SaveState != nil {
		r.SaveState()
	}
}

func (r *Runner) markFailure(revision Revision, errText string) {
	if r.State == nil {
		return
	}
	r.stateMu.Lock()
	defer r.stateMu.Unlock()
	r.State.MarkFailure(revision.Note.NoteID, errText, r.Now())
	if r.SaveState != nil {
		r.SaveState()
	}
}

func (r *Runner) uploadRevision(revision Revision) (uploadResult, error) {
	payload := MetadataPayload(r.Account, revision, r.Now())
	modifiedAt := common.ISOFormat(revision.Note.ModifiedAt)
	noteID := revision.Note.NoteID
	result := uploadResult{}
	for _, attachment := range revision.Note.Attachments {
		if attachment.IsMissing {
			result.attachmentsMissing++
		}
	}
	if !revision.Note.IsDeleted {
		html := revision.Note.BodyHTML
		if html == "" {
			html = HTMLDocumentForNote(revision.Note)
		}
		if html != "" {
			if _, err := r.Client.UploadAppleNotesBody([]byte(html), noteID, revision.RevisionID, modifiedAt); err != nil {
				return result, err
			}
			result.htmlUploaded = 1
		}
		for _, attachment := range revision.Note.Attachments {
			if attachment.IsMissing || attachment.Path == "" || attachment.ContentSHA256 == "" {
				continue
			}
			content, err := os.ReadFile(attachment.Path)
			if err != nil {
				return result, err
			}
			if _, err := r.Client.UploadAppleNotesAttachment(content, ingestclient.AppleNotesAttachment{
				NoteID: noteID, RevisionID: revision.RevisionID, ModifiedAt: modifiedAt,
				AttachmentID: attachment.AttachmentID, Filename: attachment.Filename, ContentType: attachment.ContentType,
			}); err != nil {
				return result, err
			}
			result.attachmentsUploaded++
		}
	}
	if _, err := r.Client.UploadAppleNotesRevision(payload, noteID, revision.RevisionID, modifiedAt, revision.Fingerprint); err != nil {
		return result, err
	}
	result.revisionUploaded = 1
	result.metadataUploaded = 1
	return result, nil
}

// RevisionFromNote fingerprints the note payload; the fingerprint is the
// revision id.
func RevisionFromNote(note Note) Revision {
	fingerprint := common.JSONSHA256(NotePayload(note))
	return Revision{Note: note, RevisionID: fingerprint, Fingerprint: fingerprint, IsTombstone: note.IsDeleted}
}

// NotePayload is the fingerprinted record.
func NotePayload(note Note) map[string]any {
	attachments := make([]any, 0, len(note.Attachments))
	for _, attachment := range note.Attachments {
		attachments = append(attachments, AttachmentPayload(attachment))
	}
	raw := note.Raw
	if raw == nil {
		raw = map[string]any{}
	}
	return map[string]any{
		"note_id":            note.NoteID,
		"title":              note.Title,
		"folder_id":          note.FolderID,
		"folder_path":        note.FolderPath,
		"apple_account_id":   note.AppleAccountID,
		"apple_account_name": note.AppleAccountName,
		"created_at":         common.ISOFormat(note.CreatedAt),
		"modified_at":        common.ISOFormat(note.ModifiedAt),
		"body_text":          note.BodyText,
		"body_html":          note.BodyHTML,
		"body_markdown":      note.BodyMarkdown,
		"attachments":        attachments,
		"is_deleted":         note.IsDeleted,
		"raw":                raw,
	}
}

// AttachmentPayload is one attachment's record.
func AttachmentPayload(attachment Attachment) map[string]any {
	raw := attachment.Raw
	if raw == nil {
		raw = map[string]any{}
	}
	return map[string]any{
		"attachment_id":  attachment.AttachmentID,
		"note_id":        attachment.NoteID,
		"filename":       attachment.Filename,
		"content_type":   attachment.ContentType,
		"size_bytes":     attachment.SizeBytes,
		"content_sha256": attachment.ContentSHA256,
		"is_missing":     attachment.IsMissing,
		"error":          attachment.Error,
		"raw":            raw,
	}
}

// MetadataPayload is the revision sidecar.
func MetadataPayload(account string, revision Revision, exportedAt time.Time) map[string]any {
	note := NotePayload(revision.Note)
	note["revision_id"] = revision.RevisionID
	note["content_sha256"] = revision.Fingerprint
	return map[string]any{
		"schema_version": int64(1),
		"source":         "apple_notes",
		"account":        account,
		"exported_at":    common.ISOFormat(exportedAt),
		"note":           note,
	}
}

// HTMLDocumentForNote wraps a note's body when it has no HTML of its own.
func HTMLDocumentForNote(note Note) string {
	title := note.Title
	if title == "" {
		title = note.NoteID
	}
	body := note.BodyHTML
	if body == "" {
		body = "<pre>" + PyHTMLEscape(note.BodyText) + "</pre>"
	}
	return `<!doctype html><html><head><meta charset="utf-8"><title>` + PyHTMLEscape(title) + `</title></head><body>` + body + `</body></html>`
}
